extractor/image/image.go

```diff
@@ -0,0 +1,240 @@
+		return nil
+	}
+	ctx := context.Background()
+	var err error
+	for _, transport := range img.transports {
+		imgName := transport + img.name
+		var ref imageTypes.ImageReference
+		ref, err = alltransports.ParseImageName(imgName)
+		if err != nil {
+			return xerrors.Errorf("failed to parse an image name: %w", err)
+		}
+
+		var rawSource imageTypes.ImageSource
+		rawSource, err = ref.NewImageSource(ctx, img.systemContext)
+		if err != nil {
+			// try next transport
+			continue
+		}
+
+		var src imageTypes.ImageCloser
+		src, err = image.FromSource(ctx, img.systemContext, rawSource)
+		if err != nil {
+			return xerrors.Errorf("failed to initialize: %w", err)
+		}
+
+		img.rawSource = rawSource
+		img.src = src
+		return nil
+	}
+	// return only the last error
+	return err
+}
+
+func (img *Image) LayerInfos() ([]imageTypes.BlobInfo, error) {
+	// ignore the cache error
+	layerInfos, _ := img.getLayerInfosInCache(img.name, img.isFile)
+	if layerInfos != nil {
+		return layerInfos, nil
+	}
+
+	// When it doesn't hit a cache, it fetches the image from Docker Engine or Docker Registry
+	if err := img.populateSource(); err != nil {
+		return nil, xerrors.Errorf("failed population: %w", err)
+	}
+
+	layers := img.src.LayerInfos()
+
+	// ignore the cache error
+	_ = img.storeLayerInfosInCache(img.name, img.isFile, layers)
+
+	return layers, nil
+}
+
+func (img Image) getLayerInfosInCache(imageName string, isFile bool) ([]imageTypes.BlobInfo, error) {
+	// tar file doesn't have info in cache
+	if isFile {
+		return nil, nil
+	}
+	rc := img.cache.Get("layerinfos::" + imageName)
+	if rc == nil {
+		return nil, nil
+	}
+	var layerInfos []imageTypes.BlobInfo
+	if err := json.NewDecoder(rc).Decode(&layerInfos); err != nil {
+		return nil, xerrors.Errorf("invalid cache: %w", err)
+	}
+	return layerInfos, nil
+}
+
+func (img Image) storeLayerInfosInCache(imageName string, isFile bool, layers []imageTypes.BlobInfo) error {
+	// it doesn't store the information of a tar file
+	if isFile {
+		return nil
+	}
+
+	b, err := json.Marshal(layers)
+	if err != nil {
+		return xerrors.Errorf("invalid json: %w", err)
+	}
+	return img.cache.SetBytes("layerinfos::"+imageName, b)
+}
+
+func (img *Image) ConfigBlob(ctx context.Context) ([]byte, error) {
+	b, _ := img.getConfigBlobInCache(img.name, img.isFile)
+	if b != nil {
+		return b, nil
+	}
+
+	if err := img.populateSource(); err != nil {
+		return nil, xerrors.Errorf("failed population: %w", err)
+	}
+
+	blob, err := img.src.ConfigBlob(ctx)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to get a config: %w", err)
+	}
+
+	if !img.isFile {
+		_ = img.cache.SetBytes("configblob::"+img.name, blob)
+	}
+
+	return blob, nil
+}
+
+func (img Image) getConfigBlobInCache(imageName string, isFile bool) ([]byte, error) {
+	// tar file doesn't have info in cache
+	if isFile {
+		return nil, nil
+	}
+	rc := img.cache.Get("configblob::" + imageName)
+	if rc == nil {
+		return nil, nil
+	}
+	b, err := ioutil.ReadAll(rc)
```

Comment:
How big these blobs can get? We're doing `ioutil.ReadAll()` here so should be mindful.

Reply:
It's around 1~2 KB.
```
$ ls -alh fanal/configblob::*
-rw-r--r-- 1 teppei 1.5K  1 26 14:42 fanal/configblob::docker.io_library_alpine:3.7
-rw-r--r-- 1 teppei 2.2K  1 26 14:50 fanal/configblob::docker.io_library_centos:7
-rw-r--r-- 1 teppei  591  1 26 14:50 fanal/configblob::gcr.io_distroless_base-debian10:latest
```